server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -222,42 +221,60 @@ private int findRootDocumentIfNested(SearchContext context, LeafReaderContext su
         return -1;
     }
 
-    private SearchHit createSearchHit(SearchContext context,
-                                      FieldsVisitor fieldsVisitor,
-                                      int docId,
-                                      int subDocId,
-                                      Map<String, Set<String>> storedToRequestedFields,
-                                      LeafReaderContext subReaderContext) {
+    /**
+     * Resets the provided {@link FetchSubPhase.HitContext} with information on the current
+     * document. This includes the following:
+     *   - Adding an initial {@link SearchHit} instance.
+     *   - Loading the document source and setting it on {@link SourceLookup}. This allows
+     *     fetch subphases that use the hit context to access the preloaded source.
+     */
+    private void prepareHitContext(FetchSubPhase.HitContext hitContext,
+                                   SearchContext context,
+                                   FieldsVisitor fieldsVisitor,
+                                   int docId,
+                                   int subDocId,
+                                   Map<String, Set<String>> storedToRequestedFields,
+                                   LeafReaderContext subReaderContext) {
         if (fieldsVisitor == null) {
-            return new SearchHit(docId, null, null, null);
-        }
-        loadStoredFields(context.shardTarget(), subReaderContext, fieldsVisitor, subDocId);
-        fieldsVisitor.postProcess(context.mapperService());
-        SearchHit searchHit;
-        if (fieldsVisitor.fields().isEmpty() == false) {
-            Map<String, DocumentField> docFields = new HashMap<>();
-            Map<String, DocumentField> metaFields = new HashMap<>();
-            fillDocAndMetaFields(context, fieldsVisitor, storedToRequestedFields, docFields, metaFields);
-            searchHit = new SearchHit(docId, fieldsVisitor.id(), docFields, metaFields);
+            SearchHit hit = new SearchHit(docId, null, null, null);
+            hitContext.reset(hit, subReaderContext, subDocId, context.searcher());
         } else {
-            searchHit = new SearchHit(docId, fieldsVisitor.id(), emptyMap(), emptyMap());
-        }
-        // Set _source if requested.
-        SourceLookup sourceLookup = context.lookup().source();
-        sourceLookup.setSegmentAndDocument(subReaderContext, subDocId);
-        if (fieldsVisitor.source() != null) {
-            sourceLookup.setSource(fieldsVisitor.source());
```

Comment:
I wonder if we can still assign the `context.lookup().source()` with the current _source so that `_script` can share the source with other fetch sub-phase. That would only work for sub-phases that use a script and override `hitExecute` so the new `FetchFieldsPhase` would benefit from this hack ? We can then think how we could clenup the context for script in a follow up ?

Reply:
I don't think `FetchFieldsPhase` would benefit currently, because it doesn't support scripts? I like the plan of considering it in a follow-up, for example as part of converting `ScriptFieldsPhase` to use `hitExecute`.

Reply:
Right, I was thinking preemptively of runtime scripted field but +1 to consider this in a follow up.